### PR TITLE
Refactor/new bulk

### PR DESCRIPTION
**Pull request by tjittedevries**
Eerst als draft, maar graag wel een eerste ronde reviewen!

### CODE REVIEW

File: src/features/bulk/bulk/sagas.ts [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+				break;
+			case CurrentStep.SelectFields:
+				navigate(BulkStepsToRoutes[CurrentStep.Upload]);
+				break;
+			case CurrentStep.Upload:
+				navigate(BulkStepsToRoutes[CurrentStep.Button]);
+				break;
+		}
+	} catch (e: any) {
+		console.log('back fail', e);
```
Comment (fortunato) [inline src/features/bulk/bulk/sagas.ts:33]: Catching an error and then only logging it is not very useful I'd argue. I actually don't understand what would trigger the catch looking at this. I'm sure you can do better than typing the exception as any as well :slightly_smiling_face:

File: src/features/bulk/bulk/slice.ts [hunk 2]
```diff
@@ -0,0 +1,72 @@
+import { createSlice, PayloadAction } from '@reduxjs/toolkit';
```
Comment (fortunato) [inline src/features/bulk/bulk/slice.ts:1]: I 'm not a fan of moving this away from a `store/` directory. I believe this has been a convention for many years and is instantly recognizable to developers as the place where you go for you redux stuff.

File: src/features/single/single/model.ts [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+export type CustomFileLightOrRejection = CustomFileLight & FileRejection;
+
+export interface ISingleState {
+	currentStep: CurrentStep;
+	file?: CustomFileLight;
+	metadata: MetadataGenericType;
+}
+
+export interface DMSUpload {
```
Comment (fortunato) [inline src/features/single/single/model.ts:18]: ?

**Review by fortunato — CHANGES_REQUESTED** [review]
TLDR :man_shrugging: 

No honestly I didn't have the stamina to reach it to the end :(
From what I did get through: nice work overall; looks good mostly.

May I suggest you produce smaller pull requests in the future? I know that is hard to do when refactoring and deciding where to draw a line but this is not very friendly towards the reviewer :wink: 
What makes the PR much larger than necessary is that many files that were moved have been registered as deleted and then newly created files. This is a shame IMHO. Not just in the context of this PR but also because the git history of these files are now hidden on a different path and effectively will be forever lost.

I think you can get a decent idea from the comments how I'd judge the bits I didn't get to.

File: README.md [hunk 4]
```diff
@@ -66,7 +66,7 @@ from [@amsterdam/asc-ui](https://www.npmjs.com/package/@amsterdam/asc-ui).
 
 ## Development
 
-To bootstrap the app in a static frontend served by webpack devserver run `npm run serve`.
+To bootstrap the app in a static frontend served by webpack devserver run `npm run serve` and `npm run serve:mock-api`.
```
[the remark this answers is not in the document]
Comment (tjittedevries) [reply README.md:69]: Done!

File: jest.config.js [hunk 5]
```diff
@@ -72,10 +72,10 @@ module.exports = {
 
 	// A map from regular expressions to module names that allow to stub out resources with a single module
 	moduleNameMapper: {
-		'~/tests/(.*)$': '<rootDir>/tests/$1',
+		'~/tests/(.*)$': '<rootDir>/src/tests/$1',
```
[the remark this answers is not in the document]
Comment (tjittedevries) [reply jest.config.js:75]: I'm not sure anymore, I think this was an old change. I reverted the change and everything still works.

File: src/App.tsx [hunk 6]
```diff
@@ -1,93 +1,139 @@
...
+	}, [hasFiles]);
+
+	useEffect(() => {
+		if (!session) return;
+		const metadataProperties = utils.convertDmsDynamicFormFieldsToMetadataProperty(session.dynamicFormFields);
+
+		const newSchema = createSchemaFromMetadataProps(metadataProperties);
+		setSchema(newSchema);
+
+		const newUischema = createUISchemaFromMetadataProps(metadataProperties);
```
[the remark this answers is not in the document]
Comment (tjittedevries) [reply src/App.tsx:60]: done

File: src/App.tsx [hunk 7]
```diff
@@ -1,93 +1,139 @@
...
+		setUischema(newUischema);
+
+		const fields = utils.convertDmsDynamicFormFieldsToBulkMetadataFields(session.dynamicFormFields);
```
[the remark this answers is not in the document]
Comment (tjittedevries) [reply src/App.tsx:63]: done

File: src/App.tsx [hunk 8]
```diff
@@ -1,93 +1,139 @@
...
+
+		setMetadataFields(fields);
+	}, [session]);
+
+	async function getSession() {
+		if (session) return session;
```
[the remark this answers is not in the document]
Comment (tjittedevries) [reply src/App.tsx:69]: done

File: src/App.tsx [hunk 9]
```diff
@@ -1,93 +1,139 @@
...
+		name: 'some-name',
+	};
+	const metadataForm = {
+		schema,
+		uischema,
+		data: {},
+		renderers: [],
+	};
+
+	const onClose = () => console.log(':: onClose');
```
[the remark this answers is not in the document]
Comment (tjittedevries) [reply src/App.tsx:84]: done

File: src/AppStyles.ts [hunk 10]
```diff
@@ -0,0 +1,8 @@
+import styled from 'styled-components';
+
+export const AppStyles = styled.div`
+	padding: 20px;
+	background: lightblue;
```
[the remark this answers is not in the document]
Comment (tjittedevries) [reply src/AppStyles.ts:5]: done

File: src/features/bulk/bulk/Bulk.test.tsx [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+import { render } from '../../../tests/utils/testUtils';
+import { MetadataExample } from '../../../types';
+import { asset, schema, uischema, files as filesMock, fields as fieldsMock } from './__stubs__';
+import { Props } from './types';
+import { CurrentStep } from './model';
+import { STEP1, STEP2 } from './constants';
+import { Bulk } from './Bulk';
+
+const onCloseMock = jest.fn();
+const onMetadataSubmitMock = jest.fn().mockImplementation(() => Promise.resolve());
```
[the remark this answers is not in the document]
Comment (tjittedevries) [reply src/features/bulk/bulk/Bulk.test.tsx:12]: done

File: src/features/bulk/bulk/Bulk.test.tsx [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+};
+
+afterEach(() => {
+	jest.restoreAllMocks();
+});
+
+describe('<Bulk />', () => {
+	describe('Button', () => {
+		test('is rendered', () => {
+
```
[the remark this answers is not in the document]
Comment (tjittedevries) [reply src/features/bulk/bulk/Bulk.test.tsx:48]: done

File: src/features/bulk/bulk/RouteDetect.tsx [hunk 13]
```diff
@@ -0,0 +1,22 @@
+import React, { useEffect } from 'react';
+import { useLocation } from 'react-router-dom';
+
+import { useAppDispatch, useAppSelector } from '../../hooks';
+import { getCurrentStep } from './selectors';
+import { BulkRoutesToSteps } from './constants';
+import { setCurrentStep } from './slice';
+
+function RouteDetect() {
+	const currentStep = useAppSelector(getCurrentStep);
```
[the remark this answers is not in the document]
Comment (tjittedevries) [reply src/features/bulk/bulk/RouteDetect.tsx:10]: Noted, I'll get back to you on this one.

File: src/features/bulk/bulk/__stubs__/state.ts [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+		fields[2]
+	],
+	files: [],
+};
+
+export const state: IBulkState = {
+	currentStep: 0,
+	fields,
+	files,
+};
```
[the remark this answers is not in the document]
Comment (tjittedevries) [reply src/features/bulk/bulk/__stubs__/state.ts:102]: Yes, old code from when I just started working on this whole feature. 
Over-engineered? I think it's mostly messy. I'll make it cleaner.

File: src/features/bulk/bulk/constants.ts [hunk 15]
```diff
@@ -0,0 +1,18 @@
...
+export const STEP3 = '/bulk/step3'
+
+export enum BulkStepsToRoutes {
+	'/',
+	'/bulk/step1',
+	'/bulk/step2',
+	'/bulk/step3',
+}
+
+export const BulkRoutesToSteps: { [key: string]: number } = {
```
[the remark this answers is not in the document]
Comment (tjittedevries) [reply src/features/bulk/bulk/constants.ts:13]: done

File: src/features/bulk/bulk/sagas.ts (hunk 1 above)
Comment (tjittedevries) [reply src/features/bulk/bulk/sagas.ts:33]: The old code some @todos for errors as well. What would be the prefered way to handle these errors?

File: src/features/bulk/bulk/selectors.ts [hunk 16]
```diff
@@ -0,0 +1,44 @@
+import { createSelector } from '@reduxjs/toolkit';
+import { CustomFileLight } from '../../../types';
```
[the remark this answers is not in the document]
Comment (tjittedevries) [reply src/features/bulk/bulk/selectors.ts:2]: It's called like this because some components return CustomFile from @amsterdam/bmi-component-library. The Light variant picks a few fields from that type.

File: src/features/bulk/bulk/slice.ts (hunk 2 above)
Comment (tjittedevries) [reply src/features/bulk/bulk/slice.ts:1]: I also found examples in react/redux documentation where the slices per feature are found in the feature folder.

File: src/features/bulk/bulk/utils.ts [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+import { IBulkField, IDmsDynamicFormField } from "./model";
+
+function convertStringToKey(string: string): string {
+	return string.toLowerCase().replace(' ', '_')
+}
+
+export function convertDmsDynamicFormFieldsToMetadataProperty(fields: IDmsDynamicFormField[]): MetadataProperty[] {
+	const list: MetadataProperty[] = [];
+
+	fields.forEach(field => {
```
[the remark this answers is not in the document]
Comment (tjittedevries) [reply src/features/bulk/bulk/utils.ts:11]: you're right.

File: src/features/bulk/bulk/utils.ts [hunk 18]
```diff
@@ -0,0 +1,67 @@
...
+			item.customFormat = 'creatable';
+		}
+
+		list.push(item)
+	});
+
+	return list;
+}
+
+export function convertDmsDynamicFormFieldsToBulkMetadataFields(fields: IDmsDynamicFormField[]): IBulkField[] {
```
[the remark this answers is not in the document]
Comment (tjittedevries) [reply src/features/bulk/bulk/utils.ts:35]: yes fixed as above.

File: src/features/bulk/steps/Step3.tsx [hunk 19]
```diff
@@ -0,0 +1,25 @@
...
+import { useAppSelector } from '../../hooks';
+import { STEP1 } from '../bulk/constants';
+
+import { getFiles } from '../bulk/selectors';
+import { Props } from '../bulk/types';
+import BulkWizard from '../wizard/BulkWizard';
+
+export default function Step3<T>(props: Props<T>) {
+	const files = useAppSelector(getFiles);
+	// const [isValidForm, setIsValidForm] = useState<boolean>(true);
```
[the remark this answers is not in the document]
Comment (tjittedevries) [reply src/features/bulk/steps/Step3.tsx:12]: ok

File: src/features/bulk/bulk/selectors.ts [hunk 20]
```diff
@@ -0,0 +1,44 @@
+import { createSelector } from '@reduxjs/toolkit';
```
[the remark this answers is not in the document]
Comment (tjittedevries) [reply src/features/bulk/bulk/selectors.ts:1]: Yeah, at one point I started from scratch and I had the old single/bulk and new single/bulk folders. But I understand that a git history would've been nice.

File: src/features/single/single/model.ts (hunk 3 above)
Comment (tjittedevries) [reply src/features/single/single/model.ts:18]: old code I forgot to remove.

File: db.json [hunk 21]
```diff
@@ -0,0 +1,293 @@
+{
```
[the remark this answers is not in the document]
Comment (tjittedevries) [reply db.json:1]: The json-server is only used for local development, using a typed version seems overengineering to me. Plus the use of json-server is an improvement to the previous solution. If a typed version is needed, we can add that to the backlog.

File: src/features/bulk/bulk/__stubs__/state.ts [hunk 22]
```diff
@@ -0,0 +1,102 @@
+import { CustomFileLight } from '../../../../types';
+import {
+	IBulkField,
+	IBulkFile,
+	IBulkFileMetadata,
+	IBulkState,
+} from '../model';
+
+export function makeFile(
```
[the remark this answers is not in the document]
Comment (tjittedevries) [reply src/features/bulk/bulk/__stubs__/state.ts]: removed during refactoring.

File: db.json (hunk 21 above)
[the remark this answers is not in the document]
Comment (fortunato) [reply db.json:1]: I disagree with you classifying it as over-engineering. IMO it was a choice you were able to make. In my mind it always makes sense to prefer the strictly typed option if the extra effort required seems negligible.

File: src/features/bulk/bulk/Bulk.test.tsx [hunk 23]
```diff
@@ -0,0 +1,94 @@
...
+import { STEP1, STEP2 } from './constants';
+import { Bulk } from './Bulk';
+
+const onCloseMock = jest.fn();
+const onMetadataSubmitMock = jest.fn().mockImplementation(() => Promise.resolve());
+const onFileSuccessMock = jest.fn().mockImplementation(() => Promise.resolve());
+const onFileRemoveMock = jest.fn().mockImplementation(() => Promise.resolve());
+const onCancelMock = jest.fn().mockImplementation(() => Promise.resolve());
+
+const mockData: MetadataExample = {
```
[the remark this answers is not in the document]
Comment (fortunato) [reply src/features/bulk/bulk/Bulk.test.tsx]: I believe the Jest convention is `__mocks__` for mock functions

File: db.json (hunk 21 above)
[the remark this answers is not in the document]
Comment (tjittedevries) [reply db.json:1]: The data in the mock-api is to mock an example implementation with DMS. We don't maintain those types for these in this project and would have to import those from DMS? It's just a mock server to have an endpoint for an example implementation. For me the this isn't high on the priority list of all PR feedback, I'll fix the other feedback first.

File: src/features/bulk/bulk/RouteDetect.tsx (hunk 13 above)
[the remark this answers is not in the document]
Comment (tjittedevries) [reply src/features/bulk/bulk/RouteDetect.tsx:10]: made a solution

File: src/features/bulk/bulk/slice.ts (hunk 2 above)
Comment (tjittedevries) [reply src/features/bulk/bulk/slice.ts]: made a hybrid solution. store folder in the feature.

File: src/features/bulk/bulk/slice.ts [hunk 24]
```diff
@@ -0,0 +1,72 @@
...
+			state.files = newFiles;
+		},
+		resetState: (state: IBulkState, action: PayloadAction<{ navigate: NavigateFunction }>) => initialState,
+		setCurrentStep: (state: IBulkState, action: PayloadAction<CurrentStep>) => {
+			state.currentStep = action.payload;
+		},
+		// PayloadAction<any> because issues with converting the type from uknown to something useful!!!
+		setFieldData: (state: IBulkState, action: PayloadAction<any>) => {
+			const obj = { ...action.payload };
+			const newFields = [...state.fields];
```
[the remark this answers is not in the document]
Comment (tjittedevries) [reply src/features/bulk/bulk/slice.ts]: made a reducer, moved to a function.
I'm open to hearing suggestions, I feel it could be cleaner.

**Review by trickyil2000 — CHANGES_REQUESTED** [review]
It seems like you duplicated a large amount of code. Is this on purpose? I would prefer a more DRY approach.

File: src/App.tsx [hunk 25]
```diff
@@ -1,93 +1,135 @@
...
+} from './types';
+import * as utils from './features/bulk/bulk/utils';
+import { createSchemaFromMetadataProps, createUISchemaFromMetadataProps } from './utils';
+import { AppStyles } from './AppStyles';
 
-type MetadataExample = {
-	documentDescription: string;
-	executionDate: string;
+const asset = {
+	code: '1337',
```
[the remark this answers is not in the document]
Comment (tjittedevries) [reply src/App.tsx:23]: Yes, the whole App.tsx is a sample implementation, in this case for DMS.

File: src/App.tsx [hunk 26]
```diff
@@ -1,93 +1,135 @@
...
-										}}
-										basePath={basePath}
-									/>
-								</div>
-							)}
+				<AppStyles>
+					<div>
+						<Single
+							asset={asset}
+							basePath="/documents/29/bulk-metadata"
```
[the remark this answers is not in the document]
Comment (tjittedevries) [reply src/App.tsx:99]: See previous comment on App.tsx

File: src/features/single/single/constants.ts [hunk 27]
```diff
@@ -0,0 +1,17 @@
+import { CurrentStep } from "./store/model";
+
+export const STEP0 = '/';
+export const STEP1 = '/single/step1';
+export const STEP2 = '/single/step2';
+
+export enum SingleStepsToRoutes {
+	'/',
```
[the remark this answers is not in the document]
Comment (tjittedevries) [reply src/features/single/single/constants.ts:8]: Good idea.

[the remark this answers is not in the document]
Comment (tjittedevries) [reply src/features/single/single/constants.ts:8]: It doesn't seem to work to put the variables STEP0 etc inside the enum. 
SingleStepsToRoutes is there to convert a CurrentStep (0,1,2) to a path.

File: src/features/bulk/bulk/constants.ts [hunk 28]
```diff
@@ -0,0 +1,20 @@
+import { CurrentStep } from "./store/model"
+
+export const STEP0 = '/'
+export const STEP1 = '/bulk/step1'
+export const STEP2 = '/bulk/step2'
+export const STEP3 = '/bulk/step3'
+
+export enum BulkStepsToRoutes {
+	'/',
```
[the remark this answers is not in the document]
Comment (tjittedevries) [reply src/features/bulk/bulk/constants.ts:9]: see discussion for src/features/single/single/constants.ts.
